lib/spack/spack/solver/asp.py

```diff
@@ -3462,102 +3424,46 @@ def value(self) -> str:
         return "".join(self.asp_problem)
 
 
-class CompilerParser:
-    """Parses configuration files, and builds a list of possible compilers for the solve."""
+def possible_compilers(*, configuration) -> Tuple[Set["spack.spec.Spec"], Set["spack.spec.Spec"]]:
+    result, rejected = set(), set()
 
-    def __init__(self, configuration) -> None:
-        self.compilers: Set[KnownCompiler] = set()
-        for c in spack.compilers.all_compilers_from(configuration):
-            if using_libc_compatibility() and not c_compiler_runs(c):
+    # Compilers defined in configuration
+    for c in spack.compilers.config.all_compilers_from(configuration):
+        if using_libc_compatibility() and not c_compiler_runs(c):
+            rejected.add(c)
+            try:
+                compiler = c.extra_attributes["compilers"]["c"]
                 tty.debug(
-                    f"the C compiler {c.cc} does not exist, or does not run correctly."
-                    f" The compiler {c.spec} will not be used during concretization."
+                    f"the C compiler {compiler} does not exist, or does not run correctly."
+                    f" The compiler {c} will not be used during concretization."
                 )
-                continue
+            except KeyError:
+                tty.debug(f"the spec {c} does not provide a C compiler.")
 
-            if using_libc_compatibility() and not c.default_libc:
-                warnings.warn(
-                    f"cannot detect libc from {c.spec}. The compiler will not be used "
-                    f"during concretization."
-                )
-                continue
+            continue
 
-            target = c.target if c.target != "any" else None
-            candidate = KnownCompiler(
-                spec=c.spec, os=c.operating_system, target=target, available=True, compiler_obj=c
+        if using_libc_compatibility() and not CompilerPropertyDetector(c).default_libc():
```

Comment:
there are a lot of calls to compilers here -- later we may want to consider parallelizing them up front so they don't slow down setup.  I haven't measured though... maybe they're insignificant but some of these (e.g. intel w/license server) can be slow.